test/common/http/conn_pool_grid_test.cc

```diff
@@ -439,6 +473,98 @@ TEST_F(ConnectivityGridTest, SuccessAfterBroken) {
   EXPECT_TRUE(grid_.isHttp3Broken());
 }
 
+// Test the HTTP/3 pool successfully connecting when HTTP/3 is available.
+TEST_F(ConnectivityGridWithAlternateProtocolsCacheTest, Success) {
+  addHttp3AlternateProtocol();
+  EXPECT_EQ(grid_.first(), nullptr);
+
+  EXPECT_NE(grid_.newStream(decoder_, callbacks_), nullptr);
+  EXPECT_NE(grid_.first(), nullptr);
+  EXPECT_EQ(grid_.second(), nullptr);
+
+  // onPoolReady should be passed from the pool back to the original caller.
+  ASSERT_NE(grid_.callbacks(), nullptr);
+  EXPECT_CALL(callbacks_.pool_ready_, ready());
+  grid_.callbacks()->onPoolReady(encoder_, host_, info_, absl::nullopt);
+  EXPECT_FALSE(grid_.isHttp3Broken());
+}
+
+// Test that when HTTP/3 is not available then the HTTP/3 pool is skipped.
+TEST_F(ConnectivityGridWithAlternateProtocolsCacheTest, SuccessWithoutHttp3) {
+  EXPECT_EQ(grid_.first(), nullptr);
+
+  EXPECT_LOG_CONTAINS("trace", "HTTP/3 is not available to host 'hostname', skipping.",
+                      EXPECT_NE(grid_.newStream(decoder_, callbacks_), nullptr));
+  EXPECT_NE(grid_.first(), nullptr);
+  EXPECT_NE(grid_.second(), nullptr);
+
+  // onPoolReady should be passed from the pool back to the original caller.
+  ASSERT_NE(grid_.callbacks(), nullptr);
+  EXPECT_CALL(callbacks_.pool_ready_, ready());
+  grid_.callbacks()->onPoolReady(encoder_, host_, info_, absl::nullopt);
+}
+
+// Test that when HTTP/3 is not available then the HTTP/3 pool is skipped.
+TEST_F(ConnectivityGridWithAlternateProtocolsCacheTest, SuccessWithExpiredHttp3) {
+  AlternateProtocolsCache::Origin origin("https", "hostname", 9000);
+  const std::vector<AlternateProtocolsCache::AlternateProtocol> protocols = {
+      {"h3-29", "", origin.port_}};
+  alternate_protocols_.setAlternatives(origin, protocols, simTime().monotonicTime() + Seconds(5));
+  simTime().setMonotonicTime(Seconds(10));
+
+  EXPECT_EQ(grid_.first(), nullptr);
+
+  EXPECT_LOG_CONTAINS("trace", "HTTP/3 is not available to host 'hostname', skipping.",
+                      EXPECT_NE(grid_.newStream(decoder_, callbacks_), nullptr));
+  EXPECT_NE(grid_.first(), nullptr);
+  EXPECT_NE(grid_.second(), nullptr);
+
+  // onPoolReady should be passed from the pool back to the original caller.
+  ASSERT_NE(grid_.callbacks(), nullptr);
+  EXPECT_CALL(callbacks_.pool_ready_, ready());
+  grid_.callbacks()->onPoolReady(encoder_, host_, info_, absl::nullopt);
+}
+
+// Test that when HTTP/3 is not available then the HTTP/3 pool is skipped.
```

Comment:
Should this comment discuss that we currently don't support remote Alt-Svc?

Reply:
Done.